search/executor/executor.go

```diff
@@ -19,3 +19,79 @@
 // THE SOFTWARE.
 
 package executor
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/search"
+	"github.com/m3db/m3ninx/util"
+)
+
+var (
+	errExecutorClosed = errors.New("executor is closed")
+)
+
+type newIteratorFn func(rc util.RefCount, s search.Searcher, rs index.Readers) (doc.Iterator, error)
+
+type executor struct {
+	sync.RWMutex
+	util.RefCount
+
+	newIteratorFn newIteratorFn
+	readers       index.Readers
+
+	closed bool
+}
+
+// NewExecutor returns a new Executor for executing queries.
+func NewExecutor(rs index.Readers) search.Executor {
+	return &executor{
+		RefCount:      util.NewRefCount(),
+		newIteratorFn: newIterator,
+		readers:       rs,
+	}
+}
+
+func (e *executor) Execute(q search.Query) (doc.Iterator, error) {
+	e.RLock()
+	defer e.RUnlock()
+	if e.closed {
+		return nil, errExecutorClosed
+	}
+
+	s, err := q.Searcher(e.readers)
+	if err != nil {
+		return nil, err
+	}
+
+	iter, err := e.newIteratorFn(e, s, e.readers)
+	if err != nil {
+		return nil, err
+	}
+
+	return iter, nil
+}
+
+func (e *executor) Close() error {
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return errExecutorClosed
+	}
+	e.closed = true
+	e.Unlock()
+
+	// Wait for all references to the executor to be released.
```

Comment:
don't wait here, instead force all the references to release.

Reply:
I'm not entirely sure what you mean. Require the references to be released before calling Close?